src/shared/tpm2-util.c

```diff
@@ -1395,6 +1395,84 @@ static void hash_pin(const char *pin, size_t len, TPM2B_AUTH *auth) {
         sha256_finish_ctx(&hash, auth->buffer);
 }
 
+/* Hash data into the digest.
+ *
+ * If 'extend' is true, the hashing operation starts with the existing digest hash (and the digest is
+ * required to have a hash and its size must be correct). If 'extend' is false, the digest size is
+ * initialized to the correct size for 'alg' and the hashing operation does not include any existing digest
+ * hash. If 'extend' is false and no data is provided, the digest is initialized to a zero digest.
+ *
+ * On success, the digest hash will be updated with the hashing operation result and the digest size will be
+ * correct for 'alg'.
+ *
+ * This currently only provides SHA256, so 'alg' must be TPM2_ALG_SHA256. */
+int tpm2_digest_hash_buffers(
+                TPMI_ALG_HASH alg,
+                TPM2B_DIGEST *digest,
+                const uint8_t *data[],
+                const size_t len[],
+                size_t count,
+                bool extend) {
+
+        struct sha256_ctx ctx;
+
+        assert(digest);
+        assert(data != NULL || count == 0);
+        assert(len != NULL || count == 0);
+
+        if (alg != TPM2_ALG_SHA256)
+                return log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                                       "Hash alg not supported: 0x%x", alg);
+
+        if (extend && digest->size != SHA256_DIGEST_SIZE)
+                return log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                                       "Digest size 0x%x, require 0x%x",
+                                       digest->size, (unsigned int)SHA256_DIGEST_SIZE);
+
+        /* Since we're hardcoding SHA256 (for now), we can check this at compile time. */
+        assert_cc(sizeof(digest->buffer) >= SHA256_DIGEST_SIZE);
+
+        CLEANUP_ERASE(ctx);
+
+        sha256_init_ctx(&ctx);
+
+        if (extend)
+                sha256_process_bytes(digest->buffer, digest->size, &ctx);
+        else
+                *digest = (TPM2B_DIGEST){ .size = SHA256_DIGEST_SIZE, };
+
+        for (unsigned i = 0; i < count; i++)
+                sha256_process_bytes(data[i], len[i], &ctx);
+        if (extend || count > 0)
+                sha256_finish_ctx(&ctx, digest->buffer);
+
+        return 0;
+}
+
+/* Same as above, but takes TPM2B_DIGEST[] instead of uint8_t*[]. The digests may be any size digests. */
+int tpm2_digest_hash_digests(
+                TPMI_ALG_HASH alg,
+                TPM2B_DIGEST *digest,
+                const TPM2B_DIGEST srcdigests[],
+                size_t count,
+                bool extend) {
+
+        const uint8_t **data = NULL;
+        size_t *len = NULL;
+
+        if (count > 0) {
+                data = newa(typeof(*data), count);
+                len = newa(typeof(*len), count);
```

Comment:
hmm, this could be quite large? probably not the place to use stack memory, because it could be unbounded

Reply:
updated to use malloc0, thanks